File: backend/src/test/java/edonymyeon/backend/member/domain/PasswordTest.java
package edonymyeon.backend.member.domain;

import static edonymyeon.backend.global.exception.ExceptionInformation.ALREADY_ENCODED_PASSWORD;
import static edonymyeon.backend.global.exception.ExceptionInformation.MEMBER_PASSWORD_INVALID;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import edonymyeon.backend.membber.auth.domain.SimplePasswordEncoder;
import edonymyeon.backend.global.exception.BusinessLogicException;
import edonymyeon.backend.global.exception.EdonymyeonException;
import edonymyeon.backend.membber.member.domain.Password;
import edonymyeon.backend.membber.member.domain.SocialInfo;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Test;

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
@SuppressWarnings("NonAsciiCharacters")
class PasswordTest {

    @Test
    void 비밀번호를_String으로_생성한다() {
        //given
        final String password = "pass1234!";

        //when
        //then
        Assertions.assertDoesNotThrow(() -> Password.from(password));
    }

    @Test
    void 비밀번호를_SocialType으로_생성한다() {
        //given
        final SocialInfo.SocialType socialType = SocialInfo.SocialType.KAKAO;

        //when
        //then
        Assertions.assertDoesNotThrow(() -> Password.from(socialType));
    }

    @Test
    void 비밀번호가_8글자를_미만이면_예외가_발생한다() {
        //given
        final String password = "pass12!";

        //when
        //then
        assertThatThrownBy(() -> Password.from(password))
                .isInstanceOf(EdonymyeonException.class)
                .hasMessage(MEMBER_PASSWORD_INVALID.getMessage());
    }

    @Test
    void 비밀번호가_30글자를_초과하면_예외가_발생한다() {
        //given
        StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < 31; i++) {
            stringBuilder.append(i);
        }
        final String password = stringBuilder.toString();

        //when
        //then
        assertThatThrownBy(() -> Password.from(password))
                .isInstanceOf(EdonymyeonException.class)
                .hasMessage(MEMBER_PASSWORD_INVALID.getMessage());
    }

    @Test
    void 비밀번호가_형식이_맞지않으면_예외가_발생한다() {
        //given
        final String password = "password;%/?";

        //when
        //then
        assertThatThrownBy(() -> Password.from(password))
                .isInstanceOf(EdonymyeonException.class)
                .hasMessage(MEMBER_PASSWORD_INVALID.getMessage());
    }

    @Test
    void 이미_암호화된_비밀번호는_암호화될수없다() {
        //given
        final String input = "password1234!";
        final Password password = Password.from(input);
        final SimplePasswordEncoder encoder = new SimplePasswordEncoder();
        final Password encryptedPassword = password.encrypt(encoder);
        //when
        //then
        assertThatThrownBy(() -> encryptedPassword.encrypt(encoder))
                .isInstanceOf(BusinessLogicException.class)
                .hasMessage(ALREADY_ENCODED_PASSWORD.getMessage());
    }
}

File: android/domain/src/test/kotlin/com/domain/edonymyeon/model/PasswordTest.kt
package com.domain.edonymyeon.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows

internal class PasswordTest {
    @Test
    fun `패스워드에 공백을 넣으면 제거된다`() {
        val password = Password.create("1111AAAA! ")
        assertThat(password.value).isEqualTo("1111AAAA!")
    }

    @Test
    fun `패스워드는 8글자 이상이고 알파벳 1개, 숫자 1개, 특수문자 1개를 각각 최소한 포함해야 한다`() {
        assertDoesNotThrow {
            Password.create("11234a2!")
        }
    }

    @Test
    fun `패스워드가 8글자 이상이고 알파벳을 1개이상 포함하지 않으면 예외가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            Password.create("!1231211")
        }
    }

    @Test
    fun `패스워드가 8글자 이상이고 숫자를 1개이상 포함하지 않으면 예외가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            Password.create("AAAAA!!!!")
        }
    }

    @Test
    fun `패스워드가 8글자 이상이고 특수문자를 1개이상 포함하지 않으면 예외가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            Password.create("AAAA1111")
        }
    }

    @Test
    fun `패스워드가 8글자 미만이면 예외가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            Password.create("A!A1111")
        }
    }

    @Test
    fun `패스워드가 30글자 초과이면 예외가 발생한다`() {
        assertThrows<IllegalArgumentException> {
            Password.create("A!A111A!A111A!A111A!A111A!A1111")
        }
    }

    @Test
    fun `패스워드 검증에서 8글자 이상이고 알파벳 1개, 숫자 1개, 특수문자 1개를 각각 최소한 포함하면 true를 반환한다`() {
        val result = Password.validate("11234a2!")
        assertThat(result).isEqualTo(true)
    }

    @Test
    fun `패스워드 검증에서 8글자 이상이고 알파벳을 1개이상 포함하지 않으면 false를 반환한다`() {
        val result = Password.validate("!1231211")
        assertThat(result).isEqualTo(false)
    }

    @Test
    fun `패스워드 검증에서 8글자 이상이고 숫자를 1개이상 포함하지 않으면 false를 반환한다`() {
        val result = Password.validate("AAAAA!!!!")
        assertThat(result).isEqualTo(false)
    }

    @Test
    fun `패스워드 검증에서 8글자 이상이고 특수문자를 1개이상 포함하지 않으면 false를 반환한다`() {
        val result = Password.validate("AAAA1111")
        assertThat(result).isEqualTo(false)
    }

    @Test
    fun `패스워드 검증에서 8글자 미만이면 false를 반환한다`() {
        val result = Password.validate("A!A1111")
        assertThat(result).isEqualTo(false)
    }

    @Test
    fun `패스워드 검증에서 30글자 초과이면 false를 반환한다`() {
        val result = Password.validate("A!A111A!A111A!A111A!A111A!A1111")
        assertThat(result).isEqualTo(false)
    }

    @Test
    fun `동일한 패스워드이다`() {
        val password = Password.create("1111AAA!")
        val result = password.isEqualTo("1111AAA!")
        assertThat(result).isEqualTo(true)
    }

    @Test
    fun `다른 패스워드이다`() {
        val password = Password.create("1111AAA!")
        val result = password.isEqualTo("1111AAAA!")
        assertThat(result).isEqualTo(false)
    }
}
